### ISSUE TITLE
[Epic] Behaviour for transaction handling / blockchain state

### ISSUE THREAD

**Issue by da-kami** [itchysats/itchysats#2465]
Summary what to work on ordered by priority (highest prio first):

1. `monitor`: Broadcast transaction improvements
    1. Retry mechanism upon publication failure - https://github.com/itchysats/itchysats/issues/3097
2. Better tests and coverage for `model::Cfd` and `projection::Cfd` 
    1. Fix https://github.com/itchysats/itchysats/issues/2039 and add tests
    2. Work on a better test framework for model tests https://github.com/itchysats/itchysats/issues/3098
    3. Increase the test coverage for the model logic (create ticket(s) once https://github.com/itchysats/itchysats/issues/3098 is done)
3. Fee bumping for transactions that don't get included within some time / blocks:
    1. Bumping CET - https://github.com/itchysats/itchysats/issues/3099
    2. Bumping Commit - https://github.com/itchysats/itchysats/issues/491
    3. Bumping Punish - Punish is currently not monitored, we will have to tackle this once we publish+monitor punish! Bumping for Punish is important. see https://github.com/itchysats/itchysats/issues/3160
    4. Note on bumping collaborative close: Potentially create a ticket for this if it can easily be done once the other two are done.
    5. Note on refund: I think refund should be something triggered manually and managing refund should not be automated because it is a case that is really exceptional. So, an automatic bumping feature for refund might not be necessary.
6. Un-monitor once CFD is final: https://github.com/itchysats/itchysats/issues/276
7. Evaluate if https://github.com/itchysats/itchysats/issues/206 is a lower hanging fruit after finishing work on (3.) - it might include monitoring mempool.

Other decisions based on the [detailed analysis in the comment below](https://github.com/itchysats/itchysats/issues/2465#issuecomment-1266394973):

<details>
    <summary>Decision: We don't change the monitoring logic to be driven by transactions being spent but keep the current logic of monitoring each transaction individually and accept that certain scenarios cannot be depicted.</summary>
We really <i>want</i> to change the monitoring to be driven by transactions being spent rather than monitoring each transaction individually. However, `mempool.space` is not reliable enough to be used as public infrastructure provider (and there are hardly any other public instances of `mempool.space` out there) and we don't want to become an infrastructure provider yet. We *might* be able to learn that a transaction was spent using Electrum, but if we don't know *how* it was spent the solution does not add much benefit compared to the current solution - and the added complexity does not justify a change.
</details>

**Comment by da-kami** [comment]
Instead of adapting the description for now I add thoughts as comment. Eventually the description will be updated :)

---

## Context

### Blockchain state and the database

The database records blockchain and oracle state.
The database should always reflect the correct state, a change of blockchain/oracle state should always be reflected in the database.
-> The database is the *history* of blockchain/oracle state.

Note: Oracle state is already depicted correctly in the database so we focue on the blockchain state. Note that we only consume state changes from the oracle, but don't act towards the oracle (i.e. read but no write), whereas for the blockchain we read and write (i.e. publish transactions).

Assumption: Oracel state can never retroactively change, i.e. once attested an attestation is always final.

### User Interface

Our user interface was designed in a way that does not focus on the blockchain but on "CFD terminology". 
At the same time we did reflect *some* blockchain state in the UI, that is:

- Pending[Close, Force]
	- This state is used to indicates that
	- We might not need

Users may get confused by these states because we don't explain *why* this state happened in most cases.
Example: If a user closes the CFD and afterwards it is closed that is logcal. If the user closes the CFD and then it displays "pending close" the user will have think about what that means.

At the moment we still mix CFD and blockchain terminology together in the UI - I think we should have both but potentially display them separately. 
A user that wants to know details should be able to access them to understand *why* somehing happened.

## Problem -> Solution breakdown

### Problems to solve (Monitor + Database):

1. Transaction drops out of mempool
	- Montitoring mempool
		- Requires mempool feature
	- Retry publication
	- Bumping CET (in scenario where Refund is already available)
	- Bump collab settlement (?)
2. Transaction does not get published
	1. Transaction is "broken"
	    -  Can only notify user?
	2. Transaction can currently not be published
	    - Retry mechanism
3. Transaction that was confirmed gets unconfirmed because of re-org
	=> Assumption: We don't care about CFDs that have a spend transaction with at least 3 confirmations.

### Problems to solve (User Interface)

- More confidence and feedback.
	- record a separate event-log for the UI that records what happened at what point in time; this is independent of the CfdEvent, it has a different concern which is to help a user understand what's going on.
- The user should better understand *why* something happens in the UI

**Comment by da-kami** [comment]
TODO: Explore: 

Check how lock is spent rather than monitoring specific transactions.
Once we know that lock is spent we check *how* it was spent and then record state.

=> This is more efficient, because instead of monitoring a lot of different transactions (scenarios) we just monitor lock and  potentially commit.
=> For "acting" (i.e. publishing transactions) nothing changes

**Comment by da-kami** [comment]
# Overview

There are several things we can tackle:

* `monitor` improvements
   * monitoring improvements 
   * broadcast transaction improvements
* `model::Cfd` + `projection::Cfd` improvements (resulting in UX improvements)
* UX improvements (beyond model improvements)

Below I describe my thoughts and findings and then lead into what we can actually do at the current point in time.

# `monitor`: Monitoring improvements

## Current implementation

At the moment we monitor for transaction confirmation only, i.e. once we have a certain transaction available (from a use-case perspective) we monitor immediately for all transactions / timelocks that *could* happen.

<details>
    <summary><b>⬅️ Click here</b> to see Sequence diagram depicting current solution</summary>
<img src="https://user-images.githubusercontent.com/5557790/193740931-7630c2ab-00f9-41ad-bb8a-fc54d6e01a96.png"/>
</details>

## Missing features in the current monitoring solution

Note: These missing features can be tackled for the current solution, or tackled on top of improving the overall monitoring solution (see next section). If we tackle this on top of the current solution we might write code that later has to be changed again.

The monitor currently *only* focuses on confirmations. We do not take the following scenarios into account:

<details>
    <summary>1. Monitor transactions in mempool - #206</summary>
This feature is not strictly necessary but could contribute to a better UX because we have faster feedback cycles.
This would be great for more immediate feedback on what *actions* are still available for a `Cfd`.

At the moment we might only notice a state change after a transaction was confirmed, resulting in bad UX. E.g. if the maker  triggers a manual `commit`, the `taker` will only be notified about the position being `committed` after `commit` got one confirmation.
The taker can still trigger collaborative close, but it will fail on the maker side which cannot easily be explained to the taker.
</details>

<details>
    <summary>2. Monitor and bump transactions so they don't drop out of mempool - #2047 #491</summary>
This feature is important when transaction fees in the network are high, because then it can happen that transactions with low fees drop out of mempool eventually. This is especially relevant for `CET`s because we publish them with `1 sat/vbyte` at the moment.

Breakdown:

- Monitor mempool to know that a transaction is (still) in mempool or not
	- Retry publication in case the transaction dropped (after restart; this behavior is already implemented)
	- Bumping transactions according to use case scenarios:
	    - Bump CET #2047 (based on what condition? refund timelock expiring soon? after c blocks?)
	    - Bump collaborative settlement tx in case of CET timelock expiring "soon"
	        - Note: this scenario is not likely to happen
            - Bump commit if it does not get included within x blocks; commit should be included asap to start the CET/refund timelocks, see #491
</details>

<details>
    <summary>3. Monitor transaction that was confirmed gets unconfirmed because of re-org</summary>

Assumption: We don't care about CFDs that have a spend transaction with at least 3 confirmations.

We would have to react to an already confirmed transaction not being available on chain and trigger a re-publish.
Given that it is very unlikely that a block gets dropped we don't necessarily have to implement a feature where we deal with such a scenario. 
</details>

<details>
    <summary>4. Stop monitoring transactions once a `Cfd` is final - #276</summary>
Once we are in a final state (i.e lock spent collaboratively, or commit spent) it does not make sense to keep monitoring for transactions anymore. This could be one source of our ever growing memory!
This will require extending `btsieve` to being able to "unmonitor" (i.e. remove from monitoring). The monitor should be able to decide this based on confirmation of spend transactions (once final we stop).
</details>

Additionally, the monitor cannot detect a spend transaction that is not being monitored, which means we cannot detect scenarios such as:

<details>
    <summary>5. Monitor for `CET` in a `revoke-commit` case</summary>
Currently the monitor cannot depict monitoring for a `CET` after `revoke_commit` was confirmed, because we monitor by transaction ID, and the `CET` is not known when the other party publishes a 
The assumption is, that `punish` will get published, but the system should be resilient against cases where the `CET` corresponding to the `revoked_commit` gets published.

See the `Improvement proposal for monitoring` section below for more details on how we could improve the overall monitoring to be able to depict this.
</details>

<details>
    <summary>6. "Lost" collaborative settlement transaction</summary>

I ran into this scenario when recovering from backup. If I create a backup of the ItchySats database prior to triggering a collab close on a `Cfd` my `Cfd` will never get closed after recovery because I cannot monitor for collab settlement without the transaction in the database. 
Our monitoring system should be resilient against such kind of problems.

See the `Improvement proposal for monitoring` section below for more details on how we could improve the overall monitoring to be able to depict this.
</details>

Note that (5) and (6) can only tackled if we opt for the monitor refactoring captured in section below.

## Improvement proposal for monitoring

It would be more efficient and resilient to monitor for UTXOs getting spent rather than monitoring all possible transactions all the time. Furthermore, in some scenarios as described above we are unable to monitor for a specific transaction resulting in the `Cfd` state being "stuck".
If we could know *that* `lock` was spent and derive *how* it was spent (i.e. the outspend `txid`) we could derive the scenario for our use case.
Similarly we could do the same for `commit`(and `revoked_commit`) once we know that `lock` was spent by `commit`.

### Using mempool.space for monitoring "lock spent"

Using `mempool.space`'s API instead of `Electrum` would make our monitoring significantly less complex, because one can quite easily monitor for a `vout` being spent using the [`outpend` endpoint](https://mempool.space/docs/api/rest#get-transaction-outspend) of the `mempool.space` API.

<details>
    <summary><b>⬅️ Click here</b> to see Sequence diagram depicting possible solution</summary>
<img src="https://user-images.githubusercontent.com/5557790/193791313-18e56fce-de92-481d-8087-06754310ec56.png"/>
</details>


However, `mempool.space` does not disclose their API rate limits, but they state that the API is rate limited in their [docs](https://mempool.space/docs/api/rest):

> Note that we enforce rate limits. If you exceed these limits, you will get a polite error encouraging you to run your own Mempool instance. If you repeatedly exceed the limits, you may be banned from accessing the service altogether.

They don't encourage using the public API for the purpose of "monitoring", see [this discussion](https://github.com/mempool/mempool/discussions/752).

Unless we reach out to mempool.space and discuss this, for me there is too much risk in a dependency on the public `mempool.space` API for monitoring. We should stick to `Electrum` for the moment. Sidenote: The dependency on the public `Electrum` instance is also not ensured, but blockstream appears to be less strict here; and there are others out there providing public Electrum nodes.

### Using Electrum for monitoring "lock spent"

#### Can we detect that lock was spent using Electrum? **Yes, I think**

I think with [`blockchain-scripthash-listunspent`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-scripthash-listunspent) we should be able to know if a `vout` was spent or not. 

@bonomat @luckysori we would still have to verify if my understanding of this API call is correct.

#### Can we detect *how* lock was spent using Electrum? **Not by directly querying for an outspend txid**

As far as I could see from the Electrum API there is no way do this, i.e. there is no way of following the spending path of one `vout` into another `vin` and retrieve the corresponding transactions. This could be added to Electrum, but we would have to do this ourselves which is not recommended.

### Conclusion:

Even if we don't opt for mempool.space can still opt for improving the Electrum monitoring logic if we are able to know that a UTXO because we would check for *how* `lock` (or `commit` or `revoked_commit`) was spent only once we know *that* it was spent.

This could be especially more efficient for monitoring `revoked` commits, because we don't have to monitor each revoke commit script all the time but only check for them once we know that `lock` was spent. 
Note: It is tricky to check the transaction in order because how do we decide that `collab close` will never appear? We might have to fall back on the current monitoring strategy of looking for "all possibilities" then until we find another transaction.

When `lock` was spent we can also check in order of probability, i.e.:

1. Check for collab close
2. Check for commit
8. Check for revoked commit

Once commit was spent:

1. Check for CET
2. Check for refund

Once revoked commit was spent:

1. Check for punish
2. Check for "revoked CET": Cannot handle this, because we don't know the CET yet, but the internal model logic still has to be able to close the CFD!

## `monitor`: Broadcast transaction improvements

We currently have the following assumptions: 

1. Once we have created a transaction that transaction is valid and *can* be broadcasted given the output was not spent already.
2. Once we broadcast a transaction it will succeed, i.e. we don't retry.

While (1) is a fair assumption, we have to become more resilient to failure with (2).

Scenarios where we fail to broadcast a transaction that we already handle:

1. Transaction already on chain (we ignore this and move on)

Scenarios where we fail to broadcast a transaction that we don't handle:

1. Transaction fails to be broadcasted (any other failure than the case mentioned above)

We will have to distinguish transient and permanent errors.
The monitor should take care of broadcast retry in case we run into a transient error (e.g. temporary network problem).
If we keep failing to broadcast with (some) transient errors we can consider notifying the user, because there might not be anything else we can do.

# `model` Improvements

Most of the model improvements are direct results of changes in the monitor, but in addition we could still evaluate `model::Cfd` and `projection::Cfd` to ensure that state changes on the blockchain get correctly reflected in the model. This is purely on a model level, i.e. we would have to make sure that a confirmation always gets priority in the model. We could add more unit tests to the `model::Cfd` logic, simulating an event history and asserting on the expected model outcome.
We could add similar tests to `projection::Cfd` as well to have more assurance on correct behavior there. We only have very few unit tests for the projection.
One edge case that is not covered is https://github.com/itchysats/itchysats/issues/2039 - we can take it as an opportunity and add unit tests when fixing this.

A more resilient model with faster state changes will result in a better UX as well because it will result in more / faster feedback for the user. So improving the monitor will likely reflect on the model which then reflects on the "view" of the model to the outside (i.e. the projection) which reflect what the user sees.

Here is another summary of model improvements that would result out of changing / extending the monitor:

* Fast Feedback by monitoring transactions in mempool 
   * Faster feedback on state change (not just upon confirmation(s))
   * Faster feedback on what actions are allowed
* Depict cases where lock/commit is spent but we don't know the transaction to monitor for (anymore)
    * Depict cases of `revoke-commit + CET`
    * Depict cases of "lost" collab close
    * Depends on the monitor being able to detect this (at the moment we can't because we cannot monitor for the previous `CET` but at the same time don't punish).

Note: I think it is good that we don't directly expose the events to the outside, i.e. the `CfdEvent` is something that is only known to the application internally, while the `projection` provides the view on the *current* state of the cfd. The consumer gets the view rather than the underlying model.
The way we implement the projection (e.g. always send out list of all cfds, ...) can be improved. This is out of scope here.

# UX improvements (beyond model improvements)

### User event log

For a user it might be hard to understand *why* a cfd is in a state without knowing the history of state changes, i.e. the history of "what happened". 
One thing that could help users understanding state changes would be an event log in the user interface. This event log could be derived from the cfd events in the database. It has to be evaluated if an event log derived from the internal events is sufficient. We also have to be careful not to expose a complete history, e.g. including all rollvers would result in a very long history.

**Comment by da-kami** [comment]
TODO:

- [x] Decide if we would consider using `mempool.space` for monitoring, it might require running public infrastructure if users run into rate limits
- [x] Evaluate if we can use `Electrum` to know that a UTXO was spent; test behavior of [`blockchain.scripthash.listunspent`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-scripthash-listunspent)
- [x] Decide if we want to change the monitor before adding additional features or if we want to add features that can be depicted with the current solution before changing the monitor. 
- [x] Create tickets based on decisions

---

- [x] Create ticket(s) for improving transaction broadcast resilience 
- [x] Create ticket(s) for model improvements, ensure sanity of always prioritize state upon confirmation + tests

**Comment by bonomat** [comment]
> I think with [`blockchain-scripthash-listunspent`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-scripthash-listunspent) we should be able to know if a `vout` was spent or not.
> 
> @bonomat @luckysori we would still have to verify if my understanding of this API call is correct.

This API will give you the transactions which will spend _to_ this address (script hash) and not from. Basically what you can do with this is, give electrum and address and see if someone sent money to it. I guess what you want is [blockchain.scripthash.subscribe](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-scripthash-subscribe).

**Comment by luckysori** [comment]
> I guess what you want is [blockchain.scripthash.subscribe](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-scripthash-subscribe).

How is `subscribe` supposed to be used? From the docs it seems like you would only be able to use it to:

1. Tell if _something_ has changed about the status of a UTXO.
2. Somehow compare the `status` hash output with a range of possible expected `status`es.

Number 1 seems almost useless, and not just for our use case. And number 2 is cumbersome and might not even be possible.

**Comment by da-kami** [comment]
I think `subscribe` would give us the same (or similar) information like `script_history`, but instead of polling the information it pushes it. At least this is my understanding. We can derive the script status, but as far as I understand that status does not tell me if the transaction was spent. 

I was hoping that with [`blockchain-scripthash-listunspent`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-scripthash-listunspent) we could derive that an output was spent - if we know that lock was confirmed and then the output of the corresponding address is not returned by `listunspent` anymore, can't we derive that it was spent then? 😅

**Comment by bonomat** [comment]
> Number 1 seems almost useless, and not just for our use case. And number 2 is cumbersome and might not even be possible.

You are right. It seems like this is [not useful for us](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-basics.html#status) as it only returns a hash of the information :/

**Comment by da-kami** [comment]
Note: Instead of a dependency on `mempool.space` we can also use Esplora. 
The Esplora API offers to monitor for spending a specific output as well, see: https://github.com/Blockstream/esplora/blob/master/API.md#get-txtxidoutspendvout

**Comment by bonomat** [comment]
This project is unmaintained
